### DISCUSSION TITLE
Moddable Nimble MTU size increase issue

### DISCUSSION THREAD

**Discussion by jainilsshah** [Moddable-OpenSource/moddable#1333] [Q&A]
I am using moddable Nimble in one project. I want to transfer 256/512 bytes over BLE between BLE server (custom esp32s3 hardware) and nrf connect mobile application (BLE client). 

In esp32s3 firmware, I have created one characteristic with write and notify properties. I have increased MTU size from nrf connect mobile application by using "requesting MTU" feature and set size to 256. On esp32s3 firmware, I used "CONFIG_BT_NIMBLE_ATT_PREFERRED_MTU=256" to increase MTU size to 256. On nrf connect application log, I see message "new MTU set to: 256". After increasing MTU on nrf connect, if I try to write on characteristic it is not successful and mobile app disconnects BLE connection after some seconds (may be due to timeout or not able to send packet over BLE). App log shows - Error 133(0x85): GATT ERROR and Error 22(0x16):  GATT CONN TERMINATE LOCAL HOST. 

Is there any other change or configurations required on esp32s3 firmware side to increase MTU size. Please help me suggesting required changes to fix this issue.

**Comment by phoddie** [comment]
> On esp32s3 firmware, I used "CONFIG_BT_NIMBLE_ATT_PREFERRED_MTU=256" to increase MTU size to 256

FWIW – that is already the default:

https://github.com/Moddable-OpenSource/moddable/blob/298fe72acdd53f371c66f19f93a40a43f0f95e4f/build/devices/esp32/xsProj-esp32s3/sdkconfig.defaults#L208

From your explanation, there is no obvious mistake. It looks like the MTU size has been successfully set based on the nRF Connect output you report. If you don't set the MTU size, are you able to write a characteristic?

**Reply by jainilsshah** [reply]
Thanks for the quick response and help to resolve this issue.

Yes, if I don't set/increase MTU size from nrf connect mobile application then I am able to write characteristic. In this case, I am able to write on characteristic (send command as per project requirement) and request packet size is lower than 23 bytes. But, respose packet size is 41 bytes (bigger than 23 bytes) so because of this issue, I am not able to get all response packet bytes on nrf connect mobile app. So, to get all the response bytes I have increased MTU size from nrf connect app.

Do I need to check any esp32s3 configuration in Moddable BLE?

**Reply by phoddie** [reply]
> Do I need to check any esp32s3 configuration in Moddable BLE?

There is no special Moddable SDK configuration for BLE MTU size. Any configuration would be from the ESP-IDF.

You might try enabling logging in the BLE server implementation to confirm that it is receiving the request to increase the MTU size. In $MODDABLE/modules/network/ble/nimble/modBLEServer.c change `#define LOG_GAP 0` and `#define LOG_GATT 0` to `1`. If the request to increase the MTU size is successfully handled, you should see `BLE_GAP_EVENT_MTU` traced in xsbug.

**Reply by jainilsshah** [reply]
Thanks for the response.

I have enabled GAP and GATT logs as mentioned above. I received below logs on BLE server implementation in xsbug. 

BLE_GAP_EVENT_CONNECT
BLE_GAP_EVENT_CONN_UPDATE_REQ
BLE_GAP_EVENT_CONN_UPDATE
BLE_GAP_EVENT_CONN_UPDATE_REQ
BLE_GAP_EVENT_CONN_UPDATE
BLE_GAP_EVENT_PASSKEY_ACTION
BLE_GAP_EVENT_ENC_CHANGE
BLE_GAP_EVENT_SUBSCRIBE
BLE_GAP_EVENT_MTU

Whenever I tried to change MTU size from nrf connect mobile application, I received **BLE_GAP_EVENT_MTU** trace message in xsbug so it it confirmed that BLE server receives MTU change request successfully.

Is there any issue in Moddable BLE implementation in **BLE_GAP_EVENT_MTU** request handling? Can I check Moddable MTU request handling by enabling logs? What are other ESP-IDF configurations I need to check? Please help me to resolve this issue asap.

**Reply by phoddie** [reply]
Good to see the `BLE_GAP_EVENT_MTU` being logged. The event is just a notification to report the new MTU size. No special handling is required.

It would be interesting to see that value, to confirm it matches what you set using the mobile app. You can do that by implementing the `onMTUExchanged` callback function, which is passed the new MTU size:

```js
class YOURCLASS extends BLEServer {
	// ...
	onMTUExchanged(size) {
		trace(`MTU size is now ${size}\n`);
	}
	// ...
}
```

**Reply by jainilsshah** [reply]
Thanks for the continued response.

I have added mentioned callback and print the MTU size. Whenever issue (not able to write on characteristic after MTU size update from nef connect mobile app) occurs, I am getting incorrect MTU size (0) on esp32s3.

I have debugged it further and enabled some more prints. I got incorrect MTU size sometimes in nimble_gap_event() callback when received BLE_GAP_EVENT_MTU event. Further this callback passes this to mtuExchangedEvent() and getting event->mtu.value value as 0 inside this API. I don't know whether nrf connect mobile app sends wrong MTU size or esp32s3 receives wrong MTU size. Is there any way to identifies this as I don't have BLE sniffer to check this. nimble_gap_event() is the first place at esp32s3 source code where BLE server receives MTU size change request from BLE client, correct? Is there anyway to check anything further at esp32s3?
[File.zip](https://github.com/Moddable-OpenSource/moddable/files/14998528/File.zip)

I am not using default esp32s3 sdkconfig default file what you have mentioned ([moddable/build/devices/esp32/xsProj-esp32s3/sdkconfig.defaults](https://github.com/Moddable-OpenSource/moddable/blob/298fe72acdd53f371c66f19f93a40a43f0f95e4f/build/devices/esp32/xsProj-esp32s3/sdkconfig.defaults#L208). Please find attached my project sdkconfig file.

**Reply by phoddie** [reply]
> I have increased MTU size from nrf connect mobile application by using "requesting MTU" feature and set size to 256

I have a question. Where is the "requesting MTU" feature in the nRF Connect mobile app? I don't see it in the iOS nRF Connect app.

**Reply by phoddie** [reply]
FWIW – I ran the [uart-server](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/network/ble/uart-server) example to experiment with this.

- Used ESP32-S3
- Latest Moddable SDK ([4.6.0](https://github.com/Moddable-OpenSource/moddable/releases/tag/4.6.0))
- ESP-IDF 5.1.2 (recommended)
- Ran nRF Connect and LightBlue mobile apps on iOS
- On connect, MTU size is automatically bumped up to 256 (add on `onMTUExchanged()` to the example to see that)
- Able to write strings of over 100 bytes to the "Rx" characteristic
- Able to receive strings of over 100 bytes from the "Tx" characteristic

This all seems as expected. What is different about your testbed?

**Reply by jainilsshah** [reply]
"Reqeusting MTU" feature is not available and not required for iOS nrf connect app as iOS BLE stack automatically sends MTU change request while connection establishment that is why MTU size is updated automatically at your end.

I am doing testing with android nrf connect BLE app and it provides features to update MTU size after connection.

I found this while searching for same issue. Please find below link. Can you please guide whether it is required or not?
https://github.com/espressif/esp-idf/issues/9627

I checked in my Moddable BLE implementation that it doesn't contain this changes.

**Comment by phoddie** [comment]
> I am doing testing with android nrf connect BLE app and it provides features to update MTU size after connection.

I see. It is new information that you are using Android. Perhaps NimBLE does not support updating the MTU size after the connection is established?

> I found this while searching for same issue. Please find below link. Can you please guide whether it is required or not?
https://github.com/espressif/esp-idf/issues/9627

> I checked in my Moddable BLE implementation that it doesn't contain this changes.

The link you shared does not propose a specific change. In fact, the developer who opened the issue noted, "In fact, I can not find the functions for increasing LL packet size...".

**Reply by jainilsshah** [reply]
Thanks for the update. I really appreciate your help and dedicated support, looking forward for the same in future.

I found this demo example inside esp-idf. Please find this demo (esp idf\examples\bluetooth\nimble\throughput_app\bleprph_throughput) and it has below implementation regarding LL packet size increase.

```c
#define LL_PACKET_LENGTH  251

static int blecent_gap_event(struct ble_gap_event *event, void *arg)
{
switch (event->type) {
        case BLE_GAP_EVENT_CONNECT:
        /* A new connection was established or a connection attempt failed. */
        if (event->connect.status == 0) {
            /* Connection successfully established. */
            /* XXX Set packet length in controller for better throughput */
            ESP_LOGI(tag, "Connection established ");
            rc = ble_hs_hci_util_set_data_len(event->connect.conn_handle,
                                              LL_PACKET_LENGTH, LL_PACKET_TIME);
    }
}
```

I think we need to perform this change according to this link - https://github.com/espressif/esp-idf/issues/12789

Please help me to conclude regarding MTU size and changes required to achieve it. Please confirm should I move ahead with this changes or not. I can perform this changes at my end and let you know the result.

**Reply by phoddie** [reply]
You can easily give that a try. The `BLE_GAP_EVENT_CONNECT` for the server is here:

https://github.com/Moddable-OpenSource/moddable/blob/3d6d0f86745a3f1f217540d40785f4a5fe45be8a/modules/network/ble/nimble/modBLEServer.c#L779-L791

After line 789, you can add:

```
            #define LL_PACKET_TIME            2120
            #define LL_PACKET_LENGTH          251
            ble_hs_hci_util_set_data_len(event->connect.conn_handle, LL_PACKET_LENGTH, LL_PACKET_TIME);
```

I look forward to hearing the results of your test. Good luck!

**Reply by jainilsshah** [reply]
I have tried with mentioned changes, but not able to solve this issue.

Perhaps NimBLE does not support updating the MTU size after the connection is established?
I am able to update MTU size once connection is established from nrf connect app. So, at very first time connection is established I am able to update MTU size and able to write characteristic (no issue observed), but if I perform connection-disconnection cycle and establish connection again at that time issue is observed - not able to write characteristic.

**Reply by jainilsshah** [reply]
Is it restriction in nimble that it allows only once to update MTU after esp32s3 boots up? Please check regarding this and let me know.

**Reply by phoddie** [reply]
> Is it restriction in nimble that it allows only once to update MTU after esp32s3 boots up? Please check regarding this and let me know.

This is a question about the NimBLE implementation on ESP-IDF. That implementation is provided by Espressif. This question is most appropriately asked of Espressif on their repo.

Of course, if there is something we can do in the Moddable SDK to support this feature, we are open to considering that. However, at this time there is no indication that the problem is caused by how the Moddable SDK uses NimBLE.

**Comment by jainilsshah** [comment]
Can you please do one favor for me? Can you please run uart-server demo that you have run earlier with android nrf connect app with below steps?

- Used ESP32-S3
- Latest Moddable SDK ([4.6.0](https://github.com/Moddable-OpenSource/moddable/releases/tag/4.6.0))
- ESP-IDF 5.1.2 (recommended)
- Connect ESP32-S3 with android nrf connect mobile app. Change MTU size from nrf connect mobile app to 256.
- Write on characteristic. It should work without any issue.
- Perform BLE disconnection.
- Connect ESP32-S3 with nrf connect mobile app again.
- Change MTU size to 256.
- Now, try to write on characteristic. Please check whether you are able to write on char successfully or not. I am facing issue in this step.
- On second time connection process onwards, I am not able to write on char after updating MTU size.

I am not using latest ESP-IDF and Moddable versions as old versions are already integrated in my project. So, Intention behind asking you for this testing is to check whether this scenario is working with latest ESP-IDF and Moddable version successfully or not. If this is working successfully then we can conclude that it may fix in newer versions and I may need to integrate new ESP-IDF and Moddable in my project.

Again, thanks for the efforts and appreciate your help and support here.

**Reply by mkellner** [reply]
@jainilsshah, I've followed your steps to try to reproduce the problem. Thank you for the step-by-step.

Using a Pixel 2 with Android 11, I am able to set the MTU to 256, connect and write. No issues. Disconnect BLE and repeat. No issues. Repeat. No issues.

**Reply by jainilsshah** [reply]
@mkellner Thanks a lot for your help here. I appreciate.

Can you please share your binaries so that I can flash same binaries on my setup and test? So, that we can easily conclude that this is Moddable and/or ESP-IDF old release issue.

**Reply by mkellner** [reply]
My target device is an unreleased product.

If you let me know which target platform you are using (ie. is it `esp32/esp32s3`), I can build a binary for you.

**Reply by jainilsshah** [reply]
@mkellner I am using esp32s3.

**Reply by mkellner** [reply]
This was built for the specific target `esp32/esp32s3`. As such, it needs to be connected to the debugger over the UART port, not USB.
[uart-server-esp32s3.zip](https://github.com/Moddable-OpenSource/moddable/files/15104908/uart-server-esp32s3.zip)

**Reply by jainilsshah** [reply]
Thanks for the help. I will test and share the result.

**Reply by jainilsshah** [reply]
@mkellner Is it possible for you to test the same with Moddable version 4.0 and ESP-IDF version 4.4 ?

**Reply by jainilsshah** [reply]
@mkellner I have done testing with your provided binaries and it is working as expected. No issue observed with it.

Can you please try to find information that Moddable version has fix this issue? I am using Moddable 4.0 and latest available is 4.6.0. So, it may be possible that it was fixed. If I find this info then it is too much helpful for me to conclude.

If possible please do testing with older Moddable version.

**Reply by phoddie** [reply]
>  I have done testing with your provided binaries and it is working as expected. No issue observed with it.

Excellent! You have a solution.

> Can you please try to find information that Moddable version has fix this issue?

Your question presumes that the only place that the issue could have been solved Is in the Moddable SDK. But others have observed the same or similar issues without using the Moddable SDK. So, it appears at least possible that the issues is in the ESP-IDF.

Only one file related to NimBLE changed between Moddable SDK 4.0 and 4.6: [modBLECommon.c](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/network/ble/nimble/modBLECommon.c). Those changes were part of moving to ESP-IDF 5. Perhaps you can run those changes on ESP-IDF 4.x. I don't know.

There were **many** BLE changes in the ESP-IDF between 4.4.3 (used in Moddable SDK 4.0) and 5.1.2 (used in Moddable SDK 4.6). That seems more likely to be the responsible for the improved behavior here.

Please let us know what you find. Thank you!

**Reply by jainilsshah** [reply]
Thanks for providing information.

I have checked by modifying [modBLECommon.c](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/network/ble/nimble/modBLECommon.c) file, but it doesn't resolve the issue.

Now, I have started looking into ESP-IDF version changes. Let me know if you have any comment on this.

**Reply by phoddie** [reply]
> I have checked by modifying [modBLECommon.c](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/network/ble/nimble/modBLECommon.c) file, but it doesn't resolve the issue.

Good that you tried. That said, the changes didn't look likely to be relevant to the issue.

> Now, I have started looking into ESP-IDF version changes. Let me know if you have any comment on this.

That seems like the right path. Good luck.

**Comment by phoddie** [comment]
Closing this out as it has gone dormant and the issue appears to be outside the scope of the Moddable SDK.
